domain/application/state/storage.go

```diff
@@ -78,6 +98,255 @@ VALUES ($storageToAdd.*)`, storageToAdd{})
+			if err != nil {
+				return errors.Errorf("creating storage parameters: %w", err)
+			}
+			if attachParams.filesystem != nil {
+				filesystemUUID, err := st.createFilesystem(ctx, tx, t.params.Pool, instUUID, netNodeUUID, *attachParams)
+				if err != nil {
+					return errors.Errorf("creating filesystem for storage %q for unit %q: %w", instUUID, unitUUID, err)
+				}
+				attachParams.filesystem.filesystemUUID = filesystemUUID
+				if err := st.attachFilesystemToNode(ctx, tx, netNodeUUID, *attachParams.filesystem); err != nil {
+					return errors.Errorf("attaching filesystem to storage %q for unit %q: %w", instUUID, unitUUID, err)
+				}
+			}
+			if attachParams.volume != nil {
+				volumeUUID, err := st.createVolume(ctx, tx, t.params.Pool, instUUID, netNodeUUID, *attachParams)
+				if err != nil {
+					return errors.Errorf("creating volume for storage %q for unit %q: %w", instUUID, unitUUID, err)
+				}
+				attachParams.volume.volumeUUID = volumeUUID
+				if err := st.attachVolumeToNode(ctx, tx, netNodeUUID, *attachParams.volume); err != nil {
+					return errors.Errorf("attaching volume to storage %q for unit %q: %w", instUUID, unitUUID, err)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (st *State) createUnitStorageInstance(ctx context.Context, tx *sqlair.TX, unitUUID coreunit.UUID, inst storageInstance) error {
+	insertStorageStmt, err := st.Prepare(`
+	INSERT INTO storage_instance (
+	   uuid,
+	   storage_name,
+	   storage_id,
+	   life_id,
+	   storage_pool,
+	   requested_size_mib,
+	   charm_uuid
+	) VALUES (
+	   $storageInstance.uuid,
+	   $storageInstance.storage_name,
+	   $storageInstance.storage_id,
+	   $storageInstance.life_id,
+	   $storageInstance.storage_pool,
+	   $storageInstance.requested_size_mib,
+	   $storageInstance.charm_uuid
+	)
+	`, storageInstance{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	insertStorageUnitStmt, err := st.Prepare(`
+	INSERT INTO storage_unit_owner (*) VALUES ($storageUnit.*)
+	`, storageUnit{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+	err = tx.Query(ctx, insertStorageStmt, inst).Run()
+	if err != nil {
+		return errors.Errorf("creating storage instance %q for unit %q: %w", inst.StorageUUID, unitUUID, err)
+	}
+
+	storageUnit := storageUnit{
+		StorageUUID: inst.StorageUUID,
+		UnitUUID:    unitUUID,
+	}
+	err = tx.Query(ctx, insertStorageUnitStmt, storageUnit).Run()
+	if err != nil {
+		return errors.Errorf("creating storage unit owner for storage %q and unit %q: %w", inst.StorageUUID, unitUUID, err)
+	}
+	return nil
+}
+
+// filesystemMountPoint returns a mount point to use for the given charm
+// storage. For stores with potentially multiple instances, the
+// instance ID is appended to the location.
+func filesystemMountPoint(
+	meta charmStorage,
+	storageID corestorage.ID,
+) (string, error) {
+	storageDir := paths.StorageDir(paths.OSUnixLike)
```

Comment:
This should not be done at the state layer. That's nuts.

Reply:
We need to know the storage ID to create the path. The storage ID is formulated during the txn using a sequence.